### PR TITLE
Add element-wise multiplication operator.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/3713

### CODE REVIEW

File: paddle/operators/CMakeLists.txt [hunk 1]
```diff
@@ -57,6 +57,7 @@ op_library(add_op SRCS add_op.cc add_op.cu)
 op_library(mean_op SRCS mean_op.cc mean_op.cu)
 
 op_library(mul_op SRCS mul_op.cc mul_op.cu DEPS math_function)
+op_library(element_wise_mul_op SRCS element_wise_mul_op.cc element_wise_mul_op.cu DEPS math_function)
```
Comment (qingqing01) [inline paddle/operators/CMakeLists.txt]: remove `DEPS math_function`

File: paddle/operators/element_wise_mul_op.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    auto* X = context.Input<Tensor>("X");
+    auto* Y = context.Input<Tensor>("Y");
+    auto* Z = context.Output<Tensor>("Out");
+    Z->mutable_data<T>(context.GetPlace());
+
+    auto X_e = framework::EigenVector<T>::Flatten(*X);
+    auto Y_e = framework::EigenVector<T>::Flatten(*Y);
+    auto Z_e = framework::EigenVector<T>::Flatten(*Z);
+
+    Z_e.device(context.GetEigenDevice<Place>()) = X_e * Y_e;
```
Comment (qingqing01) [inline paddle/operators/element_wise_mul_op.h]: The names of variables are all lowercas.    https://google.github.io/styleguide/cppguide.html#Variable_Names

Comment (qingqing01) [reply paddle/operators/element_wise_mul_op.h]: It's better to support `DotMulProjection` , where the second input is a row vector,
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/DotMulProjection.cpp#L57

The `broadcast` of Eigen can be used.

File: paddle/operators/CMakeLists.txt (hunk 1 above)
Comment (emailweixu) [inline paddle/operators/CMakeLists.txt]: Elementwise is one word.
elementwise_mul_op

File: paddle/operators/element_wise_mul_op.h (hunk 2 above)
Comment (emailweixu) [reply paddle/operators/element_wise_mul_op.h]: Both are needed.

File: paddle/operators/element_wise_mul_op.cc [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    x_grad->Resize(x_dims);
+    y_grad->Resize(y_dims);
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP(elemwisemul, ops::ElemWiseMulOp, ops::ElemWiseMulOpMaker,
```
Comment (emailweixu) [inline paddle/operators/element_wise_mul_op.cc]: elemwisemul => elementwise_mul

File: paddle/operators/element_wise_mul_op.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    dY->mutable_data<T>(ctx.GetPlace());
+
+    auto X_e = framework::EigenVector<T>::Flatten(*X);
+    auto Y_e = framework::EigenVector<T>::Flatten(*Y);
+    auto dX_e = framework::EigenVector<T>::Flatten(*dX);
+    auto dY_e = framework::EigenVector<T>::Flatten(*dY);
+    auto dOut_e = framework::EigenVector<T>::Flatten(*dOut);
+
+    dX_e.device(ctx.GetEigenDevice<Place>()) = dOut_e * Y_e;
+    dY_e.device(ctx.GetEigenDevice<Place>()) = X_e * dOut_e;
```
Comment (emailweixu) [inline paddle/operators/element_wise_mul_op.h]: Need to support the case where the gradient dX_e or dY_e is not needed.
@qingqing01 Most of the gradient operator need to handle the case where is one of more of the gradient of the input is not needed.

Comment (qingqing01) [reply paddle/operators/element_wise_mul_op.h]: @emailweixu Ok, I'll change the unit test framework to handle this.

File: paddle/operators/CMakeLists.txt (hunk 1 above)
Comment (gongweibao) [reply paddle/operators/CMakeLists.txt]: Thanks.
Done.

File: paddle/operators/element_wise_mul_op.cc (hunk 3 above)
Comment (gongweibao) [reply paddle/operators/element_wise_mul_op.cc]: Thanks.
Done.

File: paddle/operators/CMakeLists.txt (hunk 1 above)
Comment (gongweibao) [reply paddle/operators/CMakeLists.txt]: Thanks.
Done.

File: paddle/operators/elementwise_op.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  for (int i = 0; i < y_dims.size(); ++i) {
+    PADDLE_ENFORCE_EQ(x_dims[i + axis], y_dims[i],
+                      "Broadcast dimension mismatch.");
+    n *= y_dims[i];
+  }
+
+  for (int i = axis + y_dims.size(); i < x_dims.size(); ++i) {
+    post *= x_dims[i];
+  }
```
Comment (qingqing01) [inline paddle/operators/elementwise_op.h]: 实现来看不像是slice

File: paddle/operators/elementwise_op.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+inline void get_slice(const framework::DDim& x_dims,
+                      const framework::DDim& y_dims, const int axis, int& pre,
+                      int& n, int& post) {
```
Comment (qingqing01) [inline paddle/operators/elementwise_op.h]: 从下面实现来看，不是给x_dims, y_dims做slice得， 叫这个名字不是很理解。另外，单独放到`ementwise_op.h`里是否合适？

File: paddle/operators/elementwise_mul_op.h [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    // TODO(gongweibao): if axis is optional?
+    bool broadcast = ctx.template Attr<int>("broadcast");
+    PADDLE_ENFORCE(broadcast, "Do you forget broadcast parameter?");
+
+    int axis = ctx.template Attr<int>("axis");
+    PADDLE_ENFORCE(axis >= 0 && axis < x_dims.size(),
+                   "Axis should be in range [0, x_dims)");
+
+    int pre, n, post;
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]: 不是特别理解pre, n, post 的意思， 能够加上comments吗？ 或者更明确的name吗？

File: python/paddle/v2/framework/tests/test_elementwise_mul_op.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    def test_mul(self):
+        """ Warning
+        CPU gradient check error!
+        'X': np.random.random((32,84)).astype("float32"),
+        'Y': np.random.random((32,84)).astype("float32")
+        """
+        self.compare_grad(self.op, self.inputs)
+        self.check_grad(
+            self.op, self.inputs, ["X", "Y"], "Out", max_relative_error=0.5)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_elementwise_mul_op.py]: max_relative_error能调小一些吗？

File: paddle/operators/elementwise_mul_op.cc [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+class ElementWiseMulOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ElementWiseMulOpMaker(framework::OpProto *proto,
+                        framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The first input of elementwise mul op");
+    AddInput("Y", "The second input of elementwise mul op");
+    AddAttr<int>("axis", "Optional input parameter of elementwise mul op");
+    AddAttr<int>("broadcast", "Optional input parameter of elementwise mul op");
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/elementwise_mul_op.cc]: 从下面单测来看 axis像是y在x中的start_axis，需要更详细的说明。

File: paddle/operators/elementwise_mul_op.h [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    auto x_e = framework::EigenVector<T>::Flatten(*x);
+    auto y_e = framework::EigenVector<T>::Flatten(*y);
+    auto dx_e = framework::EigenVector<T>::Flatten(*dx);
+    auto dy_e = framework::EigenVector<T>::Flatten(*dy);
+    auto dout_e = framework::EigenVector<T>::Flatten(*dout);
+
+    if (x_dims == y_dims || product(y_dims) == 1) {
+      dx_e.device(ctx.GetEigenDevice<Place>()) = dout_e * y_e;
+      dy_e.device(ctx.GetEigenDevice<Place>()) = x_e * dout_e;
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]:
```
if (dx) dx_e.device(ctx.GetEigenDevice<Place>()) = dout_e * y_e;
if (dy) dy_e.device(ctx.GetEigenDevice<Place>()) = x_e * dout_e;
``` 
需要处理某个输入可能不用计算grad的情况

File: paddle/operators/elementwise_op.h (hunk 6 above)
Comment (qingqing01) [reply paddle/operators/elementwise_op.h]: 如果暂时只有这一个op用到，觉得先放到 `elementwise_mul_op.h` 中，后续增加新的op，再移出来都行。

File: paddle/operators/CMakeLists.txt [hunk 11]
```diff
@@ -54,11 +54,13 @@ set(DEPS_OPS
     identity_op
     minus_op
     mul_op
+    elementwise_mul_op
     recurrent_op
     scale_op)
 op_library(identity_op DEPS scale_op)
 op_library(minus_op DEPS scale_op)
 op_library(mul_op DEPS math_function)
+op_library(elementwise_mul_op SRCS elementwise_mul_op.cc elementwise_mul_op.cu)
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/CMakeLists.txt]: Done

File: paddle/operators/elementwise_op.h (hunk 6 above)
Comment (gongweibao) [reply paddle/operators/elementwise_op.h]: Done.

File: paddle/operators/elementwise_mul_op.h [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+#include "elementwise_op.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: Done.

File: paddle/operators/elementwise_mul_op.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: Done.

File: paddle/operators/elementwise_mul_op.cc [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+namespace operators {
+
+class ElementWiseMulOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto x_dim = ctx.Input<Tensor>("X")->dims();
+    auto y_dim = ctx.Input<Tensor>("Y")->dims();
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: Done.

File: paddle/operators/elementwise_mul_op.cc [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  }
+};
+
+class ElementWiseMulOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ElementWiseMulOpMaker(framework::OpProto *proto,
+                        framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The first input of elementwise mul op");
+    AddInput("Y", "The second input of elementwise mul op");
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: Done.

File: paddle/operators/elementwise_mul_op.h [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+                      "Rank of first input must >= rank of second input.")
+
+    if (x_dims == y_dims || product(y_dims) == 1) {
+      z_e.device(ctx.GetEigenDevice<Place>()) = x_e * y_e;
+      return;
+    }
+
+    // TODO(gongweibao): if axis is optional?
+    bool broadcast = ctx.template Attr<int>("broadcast");
+    PADDLE_ENFORCE(broadcast, "Do you forget broadcast parameter?");
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: broadcast和axis都是optional的参数。

File: paddle/operators/elementwise_mul_op.h (hunk 10 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: Done.

File: paddle/operators/elementwise_mul_op.h (hunk 7 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]:
```
Out = X ⊙ Y
1. shape(X) = (2, 3, 4, 5), shape(Y) = (3, 4), with axis=1
    pre=2, n=3*4,post=5
2.shape(X) = (2, 3, 4, 5), shape(Y) = (4,5)
    pre=2, n=4*5,post=1
```

Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: 已经加入到comment。

File: python/paddle/v2/framework/tests/test_elementwise_mul_op.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/v2/framework/tests/test_elementwise_mul_op.py]: Done.

File: python/paddle/v2/framework/tests/test_elementwise_mul_op.py [hunk 16]
```diff
@@ -0,0 +1,175 @@
+import unittest
+import numpy as np
+from gradient_checker import GradientChecker, create_op
+from op_test_util import OpTestMeta
+import random
+from paddle.v2.framework.op import Operator
+
+
+class TestElementwiseMulOp_Matrix(unittest.TestCase):
```
Comment (QiJune) [inline python/paddle/v2/framework/tests/test_elementwise_mul_op.py]: 需要先merge一下develop分支，因为目前的operator的测试框架有一些调整，也需要对应修改一下单元测试的写法

Comment (gongweibao) [reply python/paddle/v2/framework/tests/test_elementwise_mul_op.py]: Done.

File: paddle/operators/elementwise_mul_op.cc [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        "axis",
+        R"DOC(When broadcast is set, axis should be Y's dimension index in X.
+        )DOC")
+        .SetDefault(-1)
+        .EqualGreaterThan(-1);
+    AddAttr<int>(
+        "broadcast",
+        R"DOC(When broadcast is set, Y will be broadcast to match shape of X.
+        )DOC")
+        .SetDefault(0);
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.cc]: `broadcast` can be removed.

File: paddle/operators/elementwise_mul_op.h [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+/*
+ * Out = X ⊙ Y
+ * 1. shape(X) = (2, 3, 4, 5), shape(Y) = (3, 4), with axis=1
+ *    pre=2, n=3*4, post=5
+ * 2.shape(X) = (2, 3, 4, 5), shape(Y) = (4,5)
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]: 2. shape(X) = (2, 3, 4, 5), shape(Y) = (4,5), with axis = 2

File: paddle/operators/elementwise_mul_op.h [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+    auto y_dims = y->dims();
+    PADDLE_ENFORCE_GE(x_dims.size(), y_dims.size(),
+                      "Rank of first input must >= rank of second input.")
+
+    if (x_dims == y_dims || product(y_dims) == 1) {
+      z_e.device(ctx.GetEigenDevice<Place>()) = x_e * y_e;
+      return;
+    }
+
+    int axis = ctx.template Attr<int>("axis");
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]: `ctx.Attr<int>`

File: paddle/operators/elementwise_mul_op.cc [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+When shape(Y) does not equal shape(X),Y will be broadcasted 
+to match the shape of X and axis should be dimension index Y in X
+        )DOC")
+        .SetDefault(-1)
+        .EqualGreaterThan(-1);
+
+    AddOutput("Out", "The output of elementwise mul op");
+    AddComment(R"DOC(
+Limited elementwise multiple operator.The equation is: Out = X ⊙ Y.
+1. The shape of Y should be same of X or
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.cc]: The shape of Y should be same with X

File: paddle/operators/elementwise_mul_op.h [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    auto y_dims = y->dims();
+
+    auto* dx = ctx.Output<Tensor>(framework::GradVarName("X"));
+    auto* dy = ctx.Output<Tensor>(framework::GradVarName("Y"));
+    if (dx) {
+      dx->mutable_data<T>(ctx.GetPlace());
+    }
+    if (dy) {
+      dy->mutable_data<T>(ctx.GetPlace());
+    }
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]: short lines for line 119 - 124:

```
if (dx)  dx->mutable_data<T>(ctx.GetPlace());
if (dy)  dy->mutable_data<T>(ctx.GetPlace());
```

File: paddle/operators/elementwise_mul_op.cc [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    PADDLE_ENFORCE_GE(x_dims.size(), y_dims.size(),
+                      "Rank of first input must >= rank of second input.")
+
+    if (x_grad) {
+      x_grad->Resize(x_dims);
+    }
+
+    if (y_grad) {
+      y_grad->Resize(y_dims);
+    }
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.cc]: for shorter lines:

```cpp
if (x_grad) x_grad->Resize(x_dims);
if (y_grad) y_grad->Resize(y_dims);
```

File: paddle/operators/elementwise_mul_op.h [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+      }
+
+      if (dy) {
+        auto dy_e = framework::EigenVector<T>::Flatten(*dy);
+        dy_e.device(ctx.GetEigenDevice<Place>()) = x_e * dout_e;
+      }
+      return;
+    }
+
+    int axis = ctx.template Attr<int>("axis");
```
Comment (qingqing01) [inline paddle/operators/elementwise_mul_op.h]: `ctx.Attr<int>("axis");`

File: paddle/operators/elementwise_mul_op.h (hunk 21 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: https://google.github.io/styleguide/cppguide.html#Conditionals
这个应该不是强制的。

主要是gdb调试的时候比较郁闷，不知道if后边的语句到底是不是执行了。

File: paddle/operators/elementwise_mul_op.cc (hunk 22 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: 同上

File: paddle/operators/elementwise_mul_op.h (hunk 19 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: Done.

File: paddle/operators/elementwise_mul_op.h (hunk 23 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: Done.

File: paddle/operators/elementwise_mul_op.cc (hunk 17 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: 已经删除掉了。

File: paddle/operators/elementwise_mul_op.cc (hunk 20 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.cc]: Done.

File: paddle/operators/elementwise_mul_op.h (hunk 18 above)
Comment (gongweibao) [reply paddle/operators/elementwise_mul_op.h]: Done.

**Pull request merged**